scaaml/capture/aes/aes_capture.py

```diff
@@ -0,0 +1,181 @@
+    """
+
+    if measurements_info is None:
+        measurements_info = {
+            "trace": {
+                "type": "power",
+                "len": samples,
+            },
+        }
+    if capture_info is None:
+        capture_info = {}
+
+    # This over-writes the config file. The same happens with every call to
+    # Dataset.new_shard.
+    dataset = Dataset(
+        root_path=root_path,
+        shortname=shortname,
+        architecture=architecture,
+        implementation=implementation,
+        algorithm=algorithm,
+        version=version,
+        firmware_sha256=firmware_sha256,
+        description=description,
+        url=url,
+        examples_per_shard=examples_per_shard,
+        measurements_info=measurements_info,
+        attack_points_info=crypto_implementation.ATTACK_POINTS_INFO,
+        capture_info=capture_info)
+
+    # Generators of key-plaintext pairs for different splits.
+    crypto_algorithms = []
+
+    def add_crypto_alg(split: Literal['test', 'train', 'holdout'], keys: int,
+                       plaintexts: int, repetitions: int):
+        """Does not overwrite, safe to call multiple times.
+
+        Args:
+          split: Which split are we capturing.
+          keys: Number of different keys in this split.
+          plaintexts: Number of different plaintext captured with each key.
+          repetitions: Number of captures of each (key, plaintext) pair.
+        """
+        split_to_generator = {
+            'test': 'Validation',
+            'train': 'Training',
+            'holdout': 'Training',
+        }
+        if split not in split_to_generator.keys():
+            raise ValueError(
+                f'split must be one of {list(split_to_generator)},'
+                f' got {split}')
+        new_crypto_alg = SCryptoAlgorithm(
+            crypto_implementation=crypto_implementation,
+            purpose=split,
+            dataset=split_to_generator[split],
+            implementation=implementation,
+            algorithm=algorithm,
+            keys=keys,
+            plaintexts=plaintexts,
+            repetitions=repetitions,
+            examples_per_shard=examples_per_shard,
+            firmware_sha256=firmware_sha256,
+            full_kt_filename=Path(root_path) / dataset.slug /
+            f'{split}_key_text_pairs.txt',
+            full_progress_filename=Path(root_path) / dataset.slug /
+            f'{split}_progress_pairs.txt')
+        crypto_algorithms.append(new_crypto_alg)
+
+    if test_keys:
+        add_crypto_alg(split='test',
+                       keys=test_keys,
+                       plaintexts=test_plaintexts,
+                       repetitions=repetitions)
+    if train_keys:
+        add_crypto_alg(split='train',
+                       keys=train_keys,
+                       plaintexts=train_plaintexts,
+                       repetitions=repetitions)
+    if holdout_keys:
```

Comment:
We don't want to capture `holdout` with the same chip.
Valid combinations should be `train`, `train` and `test`, `test` (in case we already have `train`) or `holdout`